sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx

```diff
@@ -0,0 +1,245 @@
+  }
+
+  const TVector3 &dir_corr {pitch*dir_w.X() + (dirOffsets.X() - locOffsets.X()),  
+                            pitch*dir_w.Y() + (dirOffsets.Y() - locOffsets.Y()), pitch*dir_w.Z() + (dirOffsets.Z() - locOffsets.Z())};
+
+  pitch = dir_corr.Mag();
+
+  return pitch;
+}
+
+
+// Check if A contains B
+bool sbn::StubContains(const sbn::StubInfo &A, const sbn::StubInfo &B) {
+  if (A.vhit_hit->WireID().Plane != B.vhit_hit->WireID().Plane) return false;
+
+  for (const art::Ptr<recob::Hit> &h: A.hits) {
+    if (A.stub.OnCore(h->WireID())) {
+      if (h.key() == B.vhit_hit.key()) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+float sbn::StubDirectionDot(const sbn::StubInfo &A, const sbn::StubInfo &B, 
+    const geo::GeometryCore *geo,
+    const detinfo::DetectorPropertiesData &dprop) {
+
+  geo::Point_t vertex(A.stub.vtx);
+  
+  // project the vertex onto the wireplane
+  float vert_w = geo->WireCoordinate(vertex, A.vhit_hit->WireID()) * geo->WirePitch();
+  float vert_x = vertex.x();
+
+  float Ahit_w = A.vhit_hit->WireID().Wire * geo->WirePitch() - vert_w;
+  float Ahit_x = dprop.ConvertTicksToX(A.vhit_hit->PeakTime(), A.vhit_hit->WireID()) - vert_x;
+
+  float Bhit_w = B.vhit_hit->WireID().Wire * geo->WirePitch() - vert_w;
+  float Bhit_x = dprop.ConvertTicksToX(B.vhit_hit->PeakTime(), B.vhit_hit->WireID()) - vert_x;
+
+  float Amag = sqrt(Ahit_w*Ahit_w + Ahit_x*Ahit_x);
+  float Bmag = sqrt(Bhit_w*Bhit_w + Bhit_x*Bhit_x);
+
+  return(Ahit_w*Bhit_w + Ahit_x*Bhit_x) / (Amag*Bmag);
+}
+
+float sbn::StubTimeOffset(const sbn::StubInfo &A, const sbn::StubInfo &B, 
+    const detinfo::DetectorClocksData &dclock,
+    const detinfo::DetectorPropertiesData &dprop) {
+  // Convert the ticks to an X-point to undo tick-offsets between planes. And then convert to ticks.
+  return abs(dprop.ConvertTicksToX(A.vhit_hit->PeakTime(), A.vhit_hit->WireID()) - dprop.ConvertTicksToX(B.vhit_hit->PeakTime(), B.vhit_hit->WireID())) / dprop.DriftVelocity() / dclock.TPCClock().TickPeriod();
+}
+
+geo::Point_t sbn::TwoStubEndPosition(const sbn::PlaneTransform &T, const sbn::StubInfo &A, const sbn::StubInfo &B,
+    const geo::GeometryCore *geo,
+    const spacecharge::SpaceCharge *sce,
+    const detinfo::DetectorPropertiesData &dprop) {
+
+  static const bool VERBOSE = false;
+
+  // look up the wire-coordinate of both wires
+  double A_w = T.WireCoordinate(geo, A.vhit_hit->WireID());
+  double B_w = T.WireCoordinate(geo, B.vhit_hit->WireID());
+
+  // Use this to get y, z
+  double y = T.TwoPlaneToY(geo->View(A.vhit_hit->WireID()), A_w, geo->View(B.vhit_hit->WireID()), B_w);
+  double z = T.TwoPlaneToZ(geo->View(A.vhit_hit->WireID()), A_w, geo->View(B.vhit_hit->WireID()), B_w);
+
+  // just average the x-pos
+  double x = (dprop.ConvertTicksToX(A.vhit_hit->PeakTime(), A.vhit_hit->WireID()) + dprop.ConvertTicksToX(B.vhit_hit->PeakTime(), B.vhit_hit->WireID())) / 2.;
+
+  geo::Point_t pos(x, y, z);
```

Comment:
I don't really follow what you mean here.

Reply:
My comment is in two parts: about _x_ and about _y_ × _z_.

For _x_: let's make an example... hit times are about `0`, both of them. `dprop.ConvertTicksToX(0, plane)` returns `0`, doesn't it? So averaging the two you get `x = 0`. What is _x_ = 0 supposed to mean?

For _y_ × _z_: saying that there is a LArSoft alternative to the `TwoPlaneToY()`/`TwoPlaneToZ()` functions, provided that I understand correctly that the purpose of that code is to find the geometric "intersection" of two wires.

Reply:
Ok! I understand now about "x". I believe this is another assumption I have made that works in SBND but not ICARUS. I will add it to the list.

On y,z -- is there such a function? I could not find any.

Reply:
That would be [`geo::GeometryCore::WireIDsIntersect()`](https://nusoft.fnal.gov/larsoft/doxsvn/html/classgeo_1_1GeometryCore.html#a6e7e4d4d21650dfa9d9f84260465ba93).
Maybe you can take a look at its documentation and decide.

Reply:
I have understood my mistake on the first half of the comment: `dprop.ConvertTicksToX(0, plane)` does not return `0`. I thought it would return a _x_ coordinate _relative_ to the wire plane, but instead it returns an absolute coordinate. I have fixed the suggestion above to reflect that.
Given that, I think that _x_ is ok both for ICARUS and SBND.